src/main/java/net/minecraftforge/client/model/ModelLoader.java

```diff
@@ -273,6 +277,83 @@ public IModelState getDefaultState()
         {
             return ModelRotation.X0_Y0;
         }
+
+        @Override
+        public IModel retexture(Map<String, String> textures)
+        {
+            if (textures.isEmpty())
+                return this;
+            
+            List<BlockPart> elements = Lists.newArrayList(); //We have to duplicate this so we can edit it below.
+            for (BlockPart part : (List<BlockPart>)this.model.getElements())
+            {
+                elements.add(new BlockPart(part.positionFrom, part.positionTo, Maps.newHashMap(part.mapFaces), part.partRotation, part.shade));
+            }
+
+            ModelBlock neweModel = new ModelBlock(this.model.getParentLocation(), elements,
+                Maps.newHashMap(this.model.textures), this.model.isAmbientOcclusion(), this.model.isGui3d(), //New Textures man VERY IMPORTANT
+                new ItemCameraTransforms(this.model.getThirdPersonTransform(), this.model.getFirstPersonTransform(), this.model.getHeadTransform(), this.model.getInGuiTransform()));
+
+            Set<String> removed = Sets.newHashSet();
+            
+            for (Entry<String, String> e : textures.entrySet())
+            {
+                if (e.getValue() == null)
+                {
+                    removed.add(e.getKey());
+                    neweModel.textures.remove(e.getKey());
+                }
+                else
+                    neweModel.textures.put(e.getKey(), e.getValue());
+            }
+
+            // Map the model's texture references as if it was the parent of a model with the retexture map as its textures.
+            Map<String, String> remapped = Maps.newHashMap();
+            for (Entry<String, String> e : (Set<Entry<String, String>>)neweModel.textures.entrySet())
+                if (e.getValue().startsWith("#"))
+                    remapped.put(e.getKey(), e.getValue());
+            
+            int passes = 5;
```

Comment:
Like the comment says, it's so that if a model references "#blah" and "blah" is defined the textures map, it'll resolve that.
Multiple passes might not really be necessary, I suppose.